### PR TITLE
Simple Metrics Helper

### PR DESCRIPTION
**Pull request by keitwb**
This facilitates adding metrics to a pdata.Metrics in receivers.  It is
meant to be much easier and more fluent than than using pdata.Metrics
directly. All of the methods on it will return the same instance of
simple.Metrics as a return value, allowing you to chain method calls easily.

This is designed for cases where receivers are generating metrics from
scratch, where generally you will have a single datapoint per
metric/label combination.  It is not meant for high-volume appliations
where performance matters.

Also added a thread-safe version called SafeMetrics that can be
created from a normal builder with the AsSafe method.

### CODE REVIEW

File: consumer/pdatautil/metricsbuilder.go
Comment (james-bebbington) [inline consumer/pdatautil/metricsbuilder.go:59]: Wouldn't this usually be called `Build()`?

File: consumer/pdatautil/metricsbuilder.go
Comment (james-bebbington) [inline consumer/pdatautil/metricsbuilder.go:91]: Is there any case where you wouldn't want MetricSlice to be initialized? Maybe create a constructor and do this in there instead?

File: consumer/pdatautil/metricsbuilder.go
Comment (james-bebbington) [inline consumer/pdatautil/metricsbuilder.go:32]: You seem to mean one data point per metric based on the code, which I think is a severely limiting factor. It's pretty common to have multiple data points per metric isn't it?

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go:91]: `mb.metrics` would have to be initialized for this to do anything useful, but the main reason I want to avoid a constructor is so that you can simply create the struct using struct literals (as shown in the example test), which is  simpler and easier to write and read than a constructor either with positional args, the With* functional options technique, or setting struct fields individually with `=` after construction.

Also, this initialization has to be done repeatedly after each flush of the current metric set so even if you had a constructor the actual initialization code would still have to be a separate function that is reusable.

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go:32]: This is an area that I need to optimize in the builder -- by making it reuse `pdata.Metric` instances where the metric name and type are the same.  Even as it is now, you can add any number of datapoints per metric you want, but they will just be separate `pdata.Metric` instances which is kind of wasteful.

The one thing that is going to be very rare when generating metrics "from scratch" is having datapoints with the same metric name/label-set but different timestamps.  Technically you can do this with the builder as is by changing the Timestamp field and making new metrics, if somebody really needs that.

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go:59]: For somebody who is thinking in terms of Java builders (which I kind of am), `Build` is a bit misleading because of the state change that happens internally that causes subsequent calls to it to return an empty set of metrics.  This thing is like a buffer in that sense.  I also considered `Metrics()` as a very generic way of conveying that it generates the metrics without any of the connotations that Build() has about state.  What do you think about that?

**Comment by keitwb** [comment]
I'll make an issue describing the need for it and desired functionality.

File: consumer/pdatautil/metricsbuilder.go
Comment (rmfitzpatrick) [inline consumer/pdatautil/metricsbuilder.go:123]: I think adopting otel metric terminology would be preferable for the project:
`AddGauge` -> `AddInt64UpDownCounter`,
`AddGaugeD` ->`AddFloat64UpDownCounter`,
`AddCumulative` -> `AddInt64Counter`,
`AddCumulativeD` -> `AddFloat64Counter`

File: consumer/pdatautil/metricsbuilder.go
Comment (rmfitzpatrick) [reply consumer/pdatautil/metricsbuilder.go:59]: How about something like `FinalBuild` since this is a builder method?  Flush connotes transmission/writes via streams, and something like this could communicate the destructiveness of its `MoveAndAppendTo`.

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go:123]: I have never seen the term `UpDownCounter` before but it just seems like a very peculiar way to say "gauge", which has a long established usage across many monitoring systems.

Also neither of these match the metric type names in `pdata` (see `metrics.go`).  For cumulative they are called `monotonic` in pdata, but that is more of a mathematical property of them whereas "cumulative" better conveys the semantics in an observability context.

File: consumer/pdatautil/metricsbuilder.go
Comment (rmfitzpatrick) [reply consumer/pdatautil/metricsbuilder.go:32]: I think generating metrics with the same label keys and different label values for each point would be relatively common. 
 This is done this w/ per-cpu in the proposed docker stats receiver.  The model was also created for this use case afaict.

Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go:32]: I think you are thinking of the OTLP format where label keys and values are split up -- the `pdata.Metrics` format is more like our SignalFx datapoints where it is just a key/value pair all in one.  In `pdata.Metrics` you can group all metrics with the same name and type into a single `Metric` object, which is what I need to do.

File: consumer/pdatautil/metricsbuilder.go
Comment (rmfitzpatrick) [reply consumer/pdatautil/metricsbuilder.go:123]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#metric-instruments

Comment (rmfitzpatrick) [reply consumer/pdatautil/metricsbuilder.go:123]: If you wouldn't want to adopt the terminology of the specification for instruments, I'd at least use the metric types since gauge and cumulative don't appear to be adopted by otel.

File: consumer/pdatautil/metricsbuilder.go
Comment (rmfitzpatrick) [reply consumer/pdatautil/metricsbuilder.go:32]: Is pdata not a wrapper for otlp and without format https://github.com/open-telemetry/opentelemetry-collector/blob/master/consumer/pdata/generated_metrics.go#L530 ? I see this limitation is in its datapoint api though.

**Comment by bogdandrutu** [comment]
Please wait few days until the new metrics is released from proto

File: consumer/pdatautil/metricsbuilder.go
Comment (tigrannajaryan) [inline consumer/pdatautil/metricsbuilder.go:69]: Can we initialize the Attributes() map using InitEmptyWithCapacity() before the for loop?

File: consumer/pdatautil/metricsbuilder.go
Comment (tigrannajaryan) [reply consumer/pdatautil/metricsbuilder.go:123]: OTLP was updated and uses "gauge" terminology: https://github.com/open-telemetry/opentelemetry-proto/blob/313a868be259dce6c6516dd417d3ad5fd3321acf/opentelemetry/proto/metrics/v1/metrics.proto#L139

File: consumer/pdatautil/metricsbuilder.go
Comment (tigrannajaryan) [inline consumer/pdatautil/metricsbuilder.go:179]: Best to call `dps.At(0)` once. At() is fast but not zero cost.

File: consumer/pdatautil/metricsbuilder.go
Comment (tigrannajaryan) [inline consumer/pdatautil/metricsbuilder.go:199]: Please don't leave commented code.

File: consumer/pdatautil/metricsbuilder.go
Comment (tigrannajaryan) [inline consumer/pdatautil/metricsbuilder.go:104]: Do we really need the copying semantics? It is a performance hit for the WithLabels() case. Can we require the caller give up the ownership of the map and just copy the reference to the map in WithLabels().
(I don't know if performance matters for this call, I assume it may be called a lot when generating metrics, but I may be wrong and it is insignificant in which case copying is indeed safer and preferable).

File: consumer/pdatautil/metricsbuilder_test.go
Comment (tigrannajaryan) [inline consumer/pdatautil/metricsbuilder_test.go:47]: Is this indentation dictated by go fmt? (Looks a bit strange).

File: testutil/metricstestutil/metricsutil.go
Comment (tigrannajaryan) [inline testutil/metricstestutil/metricsutil.go:164]: Please add a doc comment. It is not clear from the name what it sorts.

**Review by tigrannajaryan — COMMENTED** [review]
I am a bit concerned about performance implications of using a map for labels. There is a reason we use slices for labels: maps are slower than slices.

This is likely non-issue for receivers that generate a few local metrics once every few seconds or so.

But I am worried that people will find this helper convenient and will use for receivers that get a high volume of data from the network where performance matters and there will be a performance hit because we first put labels into a map, then copy the map, then convert it to a slice during flushing.

A benchmark that compares this to a an equivalent approach without the helper using pdata functions directly would help make a decision.

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by tigrannajaryan** [comment]
@keitwb any thoughts on performance comment I posted above?

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go]: Given how many bugs I have run into in the Smart Agent that stemmed from mutating a map that was passed around to different goroutines, I have found it better to just copy maps and take a small performance hit and totally avoid the risk of map induced panics.  Since Go has no notion of ownership in the compiler (or test utilities that I'm aware of), so there is no way to programatically ensure something truly "gives up ownership" and stays that way.

File: consumer/pdatautil/metricsbuilder_test.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder_test.go]: Yes, the way Github does tabs with 8 spaces also doesn't help.

**Comment by keitwb** [comment]
> But I am worried that people will find this helper convenient and will use for receivers that get a high volume of data from the network where performance matters and there will be a performance hit because we first put labels into a map, then copy the map, then convert it to a slice during flushing.

Yes, that is not the intended use case for the helper to use in high-volume receivers that need performance.  It is meant for the bulk of receivers that we will need to port from the Smart Agent that generate maybe a few dozen metrics on a regular interval (usually around 10s or so). 

Maps are there because they are a standard part of the Go language and are very easy and intuitive to use for labels, and are reasonably performant even if not as fast as other data structures.  They are also more amenable to modification than slices of key/value tuple structs.  Given that the `InitFromMap` constructor used in the builder bypasses all of the slice validation code that would be present if you did `attrs.Insert*()`, I think some of the performance costs of using maps are mitigated by the lack of having to traverse the whole slice and do a bunch of `==` comparisons for every insert.

File: consumer/builder/metricsbuilder.go
Comment (rmfitzpatrick) [inline consumer/builder/metricsbuilder.go:126]: `AddSum()`?

**Comment by rmfitzpatrick** [comment]
Any chance you'd provide a readme or example of extended usage of these helpers?  There is one test w/ multiple flush calls but it just confirms no new metrics are produced.  Not immediately clear what the expected pattern is imo.

**Comment by james-bebbington** [comment]
I have two main concerns with this:

1. What's the reason for choosing `Flush` instead of `Build`? Is this mostly for performance reasons so you don't have to build and store a separate copy of the metric data in the builder? (I think we all agree the primary purpose of this is for tests but some people will likely end up using this for non-test code so that decision is understandable I guess). How about just returning an error if `Build` is called twice? While that doesn't follow the builder pattern to the letter, I would personally find that more intuitive than the current behaviour (and it would simplify the implementation).

2. The work to maintain state in the builder so you can associate repeated values against the same metric name with the same metric object makes the API not particularly intuitive imo. Why not implement nested "builders" instead so you can do something like:

```go
rmb.
    AddIntSumMetric("system.disk.io", func(mb IntSumMetricBuilder) {
        mb.
            AddDataPoint(now, 50, func(dpb IntDataPointBuilder) {
                dpb.
                    AddLabel("device", "C:").
                    AddLabel("direction", "read")
            }.
            AddDataPoint(now, 45, ...
    }).
    AddDoubleGaugeMetric(...
```

File: consumer/builder/metricsbuilder.go
Comment (keitwb) [reply consumer/builder/metricsbuilder.go:126]: Yes, will change to that.

**Comment by keitwb** [comment]
Ok, I redid the builder in a different way that gets rid of the Build method altogether and just adds metrics to a preexisting pdata.Metrics instance.  This should be much easier to share a common Metrics instance between the different instances created when you call WithLabels, which also eliminates issues with state in the builder.

**Comment by keitwb** [comment]
>     1. (I think we all agree the primary purpose of this is for tests but some people will likely end up using this for non-test code so that decision is understandable I guess).

No, exactly the opposite, it is meant for receiver code that is generating metrics from scratch, not really for tests (although it could be useful in tests).

File: consumer/pdatautil/metricsbuilder.go
Comment (keitwb) [reply consumer/pdatautil/metricsbuilder.go]: The builder now uses the new standard terms in the final Metrics model.

**Comment by james-bebbington** [comment]
> Ok, I redid the builder in a different way that gets rid of the Build method altogether and just adds metrics to a preexisting pdata.Metrics instance. This should be much easier to share a common Metrics instance between the different instances created when you call WithLabels, which also eliminates issues with state in

I guess I'm still confused by the intended interface. This code seems strange to me:

```go
var wg sync.WaitGroup
for _, disk := range []string{"sda", "sdb", "sdc"} {
	wg.Add(1)
	go func(disk string) {
		// All metrics added after this will have these labels
		diskBuilder := mb.WithLabels(map[string]string{
			"disk": disk,
		}).AddGauge("disk.usage", 9000000)

		// Add metrics in a chained manner
		diskBuilder.
			AddGauge("disk.capacity", 9000000).
			AddSum("disk.reads", 50)

		// Or add them on their own
		diskBuilder.AddGaugeD("disk.temp", 30.5)

		wg.Done()
	}(disk)
}
```

What is the use case for that workflow?
- You want to add a bunch of metrics with the same set of labels / data points
- You want to add data points for the set of common labels at the same time

Would this lead to non intuitive code if people want to add a number of metrics that don't share the set of labels & values?

Also, instead of:
```go
[ disk.reads: { points: [ { labels: { "disk": "sda" }, value: 50 }, { labels: { "disk": "sdb" }, value: 50 }, ... ] } ]
```
You're going to end up with:
```go
[ disk.reads: { points: [ { labels: { "disk": "sda" }, value: 50 } ] } ], [ disk.reads: { points: [ { labels: { "disk": "sdb" }, value: 50 } ] } ], ... ]
```
Which is a less efficient format for storing the metric in, and potentially lends itself to confusing people imo.

Wouldn't this code be much more flexible if your interface included the ability to directly add multiple data points with specific labels to a metric?

If you want the convenience of only having to specify labels once in the case where you are adding a number of metrics with the same labels, you could do something like (building on my example above):

```go
idpb := func(dpb IntDataPointBuilder) {
    dpb.
        AddLabel("device", "C:").
        AddLabel("direction", "read")
}

rmb.
    AddIntGaugeMetric("disk.capacity", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 9000000, idpb) }).
    AddIntSumMetric("disk.reads", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 50, idpb) }).
    AddDoubleGaugeMetric("disk.temp", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 30.5, idpb) }).
    ...
```

> No, exactly the opposite, it is meant for receiver code that is generating metrics from scratch, not really for tests (although it could be useful in tests).

If that's the case, then I would be a little more concerned. As per what Tigran said earlier, people would likely end up copy/pasting this code into other more performance sensitive receivers / processors. For receivers that are only creating a few metrics, is it too inconvenient to just create the metrics normally?

I was envisioning this to be used in tests as I've seen a few components implementing something similar to create lots of test data easily.

**Comment by keitwb** [comment]
> What is the use case for that workflow?

It is the first thing you listed, for adding a bunch of datapoints that share a common set of labels/dimensions.  If you look at the code in [SignalFx monitors](https://github.com/signalfx/signalfx-agent/tree/master/pkg/monitors) and [Telegraf Inputs](https://github.com/influxdata/telegraf/tree/master/plugins/inputs) it will be clear that that is by far the most common pattern used due to how things are usually processed.  I can't really think of any scenarios where you have a bunch of metrics with the same name but different labels that you are adding at the same time.  If such a thing was needed, this builder could just include more methods that accept a map of labels as map[string]string.

> is it too inconvenient to just create the metrics normally?

It is -- the pdata interface is very unfriendly to use and read.  It was designed to act as an abstraction layer between OTLP protobuf changes and collector code to facilitate specific types of changes to the protobuf model (stuff that doesn't involve adding/removing fields or fundamentally altering their type), at the cost of usability.  There is a fixed overhead embodied in the `getMetricsSlice` method that is there no matter how many metrics you have, and also quite a bit of verbosity to add each datapoint (the stuff in the switch statement cases in `addMetric`).  What has happened so far with all of the receivers that use metrics in the use case I'm describing is to make a bunch of helper methods to ease the pain and verbosity.  This is an attempt at unifying that into a single type.

> Would this lead to non intuitive code if people want to add a number of metrics that don't share the set of labels & values?

This isn't really the intended use case to add a bunch of metrics with the same name but disparate labels at the same time.  Even if you did, you just call WithLabels to specify the different labels.  Like imagine you were getting a bunch of disk metrics, you would have a loop that made a builder with a common set of labels identifying the disk, and then add your metrics to it.  That is usually the extent of the complexity of it.  You wouldn't get all the disk read stats for all the disks, and then all of the disk write stats, etc, you would get all the stats for one disk, then another, etc.  This is the case for almost everything we have come across.

> people would likely end up copy/pasting this code into other more performance sensitive receivers / processors

That is something we can spot in code review.  Also unless the metrics are already grouped by common labels it is going to be pretty hard to use the builder as it is right now.  Also even if they did basically the only performance difference that is inevitable is the conversion of maps to AttributeMap and the lack of preallocation of slices.

> Which is a less efficient format for storing the metric in, and potentially lends itself to confusing people imo.

Yes, that is the case in the current implementation but it is not inevitable and it can be solved by adding a lookup to existing metrics to append datapoints to while still maintaining constant time metric adds, or at worst just iterating the metrics to find the same metric name/type at the cost of O(n) adds.  I would consider that a performance enhancement and it doesn't change the interface at all.  Also imagine how you would do it without the help of the builder where you had metric generation done in disparate functions (you would either end up writing basically a mini builder object or passing a bunch of stuff as function parameters, or else just returning a bunch of pdata.Metric instances and then appending them to a MetricsSlice and ending up with the same issue of inefficient metric storage -- rarely would you add all of the metrics with the same name at once, usually you will be adding them grouped by the common labels due to how metrics are usually generated from service data).


```
idpb := func(dpb IntDataPointBuilder) {
    dpb.
        AddLabel("device", "C:").
        AddLabel("direction", "read")
}

rmb.
    AddIntGaugeMetric("disk.capacity", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 9000000, idpb) }).
    AddIntSumMetric("disk.reads", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 50, idpb) }).
    AddDoubleGaugeMetric("disk.temp", func(mb IntSumMetricBuilder) { mb.AddDataPoint(now, 30.5, idpb) }).
    ...
```

compared to:

```go
builder.MetricBuilder{
  Metrics: metrics,
  BaseLabels: map[string]string{
    "device": "C:",
    "direction": "read",
  },
}.AddGauge("disk.capacity", 9000000).
  AddSum("disk.reads", 50).
  AddGaugeF("disk.temp", 30.5)
```

is just so much simpler.  Without type inferencing on anonymous functions, nested functions with parameters in Go are pretty ugly (once you add the package prefix to *MetricBuilder it gets even longer).  But almost never will you need to add a bunch of metrics with the same name but different labels at once, so the flexibility is overkill.

**Comment by james-bebbington** [comment]
Okay I feel like I understand your goal a bit better now. It's different than what I initially thought which was that this was just adding wrapper / builder functions to simplify the creation of `pdata` metrics, primarily for test code.

Iiuc, for a metric with label values that do differ, which I still feel needs to be supported by this API even if it's not your primary use case (as I don't think it's that uncommon?), you would expect users to write something along the lines of:

```go
mb := builder.MetricBuilder{ Metrics: pdata.NewMetrics(), ... }

for _, disk := range []string{"sda", "sdb", "sdc"} {
    diskBuilder := mb.WithLabels(map[string]string{ "disk": disk })

    mbRead = diskBuilder.WithLabels("direction", "read")
    mbRead.AddGauge("disk.io", 1).
        AddSum("disk.ops", 2).
        AddGaugeD("disk.time", 3.5)

    mbWrite =diskBuilder.WithLabels("direction", "write")
    mbWrite.AddGauge("disk.io", 7).
        AddSum("disk.ops", 8).
        AddGaugeD("disk.time", 9.5)
}

return mb.Metrics
```

Other high level thoughts:

- It would be useful to add a test case to demonstrate this kind of usage
- I think the `Add...` functions should be called `Add...DataPoint` instead as that's really what they're doing (adding a single data point at a time)
- I think this deviates from the `Builder` pattern so maybe another term would be better?
- What is the advantage of extralabels field over just putting the extra labels into BaseLabels?
- Do we really need the safe builder? This is not for performance sensitive scenarios so why would you want to build metrics in parallel?
- Aside: My personal preference would still be for a more explicit builder where the user specifies which labels are included for each data point

> Without type inferencing on anonymous functions, nested functions with parameters in Go are pretty ugly

Yea I definitely agree with that - what I suggested is pretty ugly and probably not very idiomatic. You could implement a similar solution without the nested functions though, e.g.

```go
diskIOBuilder := NewIntGaugeMetricBuilder().SetName( "disk.io")
diskOpsBuilder := NewIntSumMetricBuilder().SetName( "disk.ops")
diskTimeBuilder := NewDoubleGaugeMetricBuilder().SetName( "disk.time")

for _, disk := range []string{"sda", "sdb", "sdc"} {
    diskLabels := NewLabelsBuilder().AddLabel("disk", disk)
    diskReadLabels := NewLabelsBuilder(diskLabels).AddLabel("direction", "read").Build()
    diskWriteLabels := NewLabelsBuilder(diskLabels).AddLabel("direction", "write").Build()

    diskIOBuilder.
        AddDataPoint(now, 1, diskReadLabels).
        AddDataPoint(now, 7, diskWriteLabels)

    diskOpsBuilder.
        AddDataPoint(now, 2, readLabels).
        AddDataPoint(now, 7, diskWriteLabels)

    diskTimeBuilder.
        AddDataPoint(now, 3, readLabels).
        AddDataPoint(now, 7, diskWriteLabels)
}

NewMetricsBuilder().
    ...
    AddMetric(diskIOBuilder.Build()).
    AddMetric(diskOpsBuilder.Build()).
    AddMetric(diskTimeBuilder.Build()).
    Build()
)
```

> Also imagine how you would do it without the help of the builder where you had metric generation done in disparate functions (you would either end up writing basically a mini builder object or passing a bunch of stuff as function parameters, or else just returning a bunch of pdata.Metric instances and then appending them to a MetricsSlice and ending up with the same issue of inefficient metric storage -- rarely would you add all of the metrics with the same name at once, usually you will be adding them grouped by the common labels due to how metrics are usually generated from service data).

What do you think of the existing code that builds metrics directly, e.g. https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/hostmetricsreceiver/internal/scraper/diskscraper/disk_scraper_others.go

**Comment by keitwb** [comment]
>     It would be useful to add a test case to demonstrate this kind of usage
Added

>     I think the Add... functions should be called Add...DataPoint instead as that's really what they're doing (adding a single data point at a time)
Changed to that format.

>     I think this deviates from the Builder pattern so maybe another term would be better?
Yes, changed it to be in a package called `simple` and the struct type just called `Metrics`, so it will appear as `simple.Metrics` in code.

>     What is the advantage of extralabels field over just putting the extra labels into BaseLabels?
It was a relic from a previous implementation so I removed it.  I don't think it had any remaining value.

>     Do we really need the safe builder? This is not for performance sensitive scenarios so why would you want to build metrics in parallel?
It isn't going to be commonly used but it is mainly there to 1) make it obvious that the base type is not thread safe, and 2) provide a convenient and simple implementation for when it is needed.  It is never really strictly necessary as you can always serialize metric generation with channels (or external mutexes) but it's just a convenience that is pretty low-effort.

**Comment by keitwb** [comment]
> What do you think of the existing code that builds metrics directly, e.g. https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/hostmetricsreceiver/internal/scraper/diskscraper/disk_scraper_others.go
It makes it harder to read than is really necessary given the relatively simple logic when all the boilerplate is removed.

File: consumer/simple/metrics_test.go
Comment (rmfitzpatrick) [inline consumer/simple/metrics_test.go:67]: Would you be able to show the expected usage w/ the next consumer's `ConsumeMetrics()` and how this could work over an interval?  Are we to be creating a scratch `Metrics{}` with every pass or would an exported `Clone()` help?

Comment (keitwb) [reply consumer/simple/metrics_test.go:67]: ConsumeMetrics just takes `metrics` as an argument.

Yes you would need to make a new builder and pdata.Metrics instance each time as sharing metric data between calls to ConsumeMetrics is forbidden.

Comment (rmfitzpatrick) [reply consumer/simple/metrics_test.go:67]: Thanks, mind adding equivalent remarks to this example for clarity?

**Review by rmfitzpatrick — APPROVED** [review]

**Comment by keitwb** [comment]
@tigrannajaryan I posted the benchmarks you requested a while back.  What else does this need to be merged?

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@keitwb Sorry, I missed the benchmarks, thanks for posting them.

@open-telemetry/collector-approvers this changes looks good to me. Benchmarks show that performance difference is very small and usability is much better. I am inclined to merge this unless anybody objects.

**Comment by tigrannajaryan** [comment]
@keitwb can you retrigger EasyCLA?

**Comment by keitwb** [comment]
@tigrannajaryan ok rebased with latest master and it passed EasyCLA.

**Comment by tigrannajaryan** [comment]
Didn't see objections, merging.

**Pull request merged**
